File: frontend/src/lib/api/valuation.ts
import { api } from './client';

export interface MethodResult {
  name: string;
  value_cents: number;
  weight: number;
  description?: string;
}

export type ValuationVerdict = 'UNDERVALUED' | 'FAIRLY_VALUED_LOW' | 'FAIR_VALUE' | 'OVERVALUED';

export type TrapFlag =
  | 'DECLINING_REVENUE'
  | 'SHRINKING_MARGINS'
  | 'GROWING_DEBT'
  | 'DECLINING_FCF'
  | 'DECLINING_EARNINGS';

export interface FundamentalTrend {
  is_value_trap: boolean;
  trap_score: number;
  flags: TrapFlag[];
  reasons: string[];
  revenue_cagr: number;
  earnings_cagr: number;
  fcf_trend: string;
  debt_trend: string;
}

export interface ValuationResult {
  ticker: string;
  current_price_cents: number;
  intrinsic_value_cents: number;
  margin_of_safety: number;
  verdict: ValuationVerdict;
  quality_score: number;
  methods: MethodResult[];
  trend?: FundamentalTrend;
}

export interface Fundamentals {
  stock_id: number;
  fiscal_year: number;
  revenue_cents: number;
  net_income_cents: number;
  eps_cents: number;
  book_value_cents: number;
  free_cash_flow_cents: number;
  dividends_cents: number;
  shares_outstanding: number;
  debt_cents: number;
  cash_cents: number;
  roe: number;
  profit_margin: number;
}

export async function getValuation(ticker: string): Promise<ValuationResult> {
  return api<ValuationResult>(`/valuation/${encodeURIComponent(ticker)}`);
}

export async function getFundamentals(ticker: string): Promise<Fundamentals> {
  return api<Fundamentals>(`/valuation/${encodeURIComponent(ticker)}/fundamentals`);
}

export async function screenWatchlist(watchlistId: number): Promise<ValuationResult[]> {
  return api<ValuationResult[]>(`/watchlists/${watchlistId}/screen`);
}

export interface BacktestResult {
  ticker: string;
  entry_date: string;
  entry_price_cents: number;
  current_price_cents: number;
  return_pct: number;
  annualized_return_pct: number;
  holding_days: number;
  entry_valuation: ValuationResult;
  model_correct: boolean;
  model_verdict: string;
  actual_outcome: string;
}

export async function backtestValuation(ticker: string, date: string): Promise<BacktestResult> {
  return api<BacktestResult>(
    `/valuation/${encodeURIComponent(ticker)}/backtest?date=${encodeURIComponent(date)}`
  );
}

File: backend/internal/domain/valuation.go
package domain

type ValuationResult struct {
	Ticker         string           `json:"ticker"`
	CurrentCents   int64            `json:"current_price_cents"`
	IntrinsicCents int64            `json:"intrinsic_value_cents"`
	MarginOfSafety float64          `json:"margin_of_safety"`
	Verdict        ValuationVerdict `json:"verdict"`
	QualityScore   int              `json:"quality_score"`
	Methods        []MethodResult   `json:"methods"`
	Trend          *FundamentalTrend `json:"trend,omitempty"`
}

type ValuationVerdict string

const (
	VerdictUndervalued ValuationVerdict = "UNDERVALUED"
	VerdictFairLow     ValuationVerdict = "FAIRLY_VALUED_LOW"
	VerdictFair        ValuationVerdict = "FAIR_VALUE"
	VerdictOvervalued  ValuationVerdict = "OVERVALUED"
)

type MethodResult struct {
	Name        string  `json:"name"`
	ValueCents  int64   `json:"value_cents"`
	Weight      float64 `json:"weight"`
	Description string  `json:"description,omitempty"`
}

type Fundamentals struct {
	StockID           int64   `json:"stock_id"`
	FiscalYear        int     `json:"fiscal_year"`
	RevenueCents      int64   `json:"revenue_cents"`
	NetIncomeCents    int64   `json:"net_income_cents"`
	EPSCents          int64   `json:"eps_cents"`
	BookValueCents    int64   `json:"book_value_cents"`
	FreeCashFlowCents int64   `json:"free_cash_flow_cents"`
	DividendsCents    int64   `json:"dividends_cents"`
	SharesOutstanding int64   `json:"shares_outstanding"`
	DebtCents         int64   `json:"debt_cents"`
	CashCents         int64   `json:"cash_cents"`
	ROE               float64 `json:"roe"`
	ProfitMargin      float64 `json:"profit_margin"`
	// Historical net income (most recent first) for earnings growth calculation.
	NetIncomeHistory []int64   `json:"net_income_history,omitempty"`
	RevenueHistory   []int64   `json:"revenue_history,omitempty"`
	FCFHistory       []int64   `json:"fcf_history,omitempty"`
	DebtHistory      []int64   `json:"debt_history,omitempty"`
	MarginHistory    []float64 `json:"margin_history,omitempty"`
}

type TrapFlag string

const (
	TrapDecliningRevenue  TrapFlag = "DECLINING_REVENUE"
	TrapShrinkingMargins  TrapFlag = "SHRINKING_MARGINS"
	TrapGrowingDebt       TrapFlag = "GROWING_DEBT"
	TrapDecliningFCF      TrapFlag = "DECLINING_FCF"
	TrapDecliningEarnings TrapFlag = "DECLINING_EARNINGS"
)

type FundamentalTrend struct {
	IsValueTrap  bool       `json:"is_value_trap"`
	TrapScore    int        `json:"trap_score"`
	Flags        []TrapFlag `json:"flags"`
	Reasons      []string   `json:"reasons"`
	RevenueCAGR  float64    `json:"revenue_cagr"`
	EarningsCAGR float64    `json:"earnings_cagr"`
	FCFTrend     string     `json:"fcf_trend"`
	DebtTrend    string     `json:"debt_trend"`
}

type BacktestResult struct {
	Ticker          string           `json:"ticker"`
	EntryDate       string           `json:"entry_date"`
	EntryPriceCents int64            `json:"entry_price_cents"`
	CurrentCents    int64            `json:"current_price_cents"`
	ReturnPct       float64          `json:"return_pct"`
	AnnualizedPct   float64          `json:"annualized_return_pct"`
	HoldingDays     int              `json:"holding_days"`
	EntryValuation  *ValuationResult `json:"entry_valuation"`
	ModelCorrect    bool             `json:"model_correct"`
	ModelVerdict    string           `json:"model_verdict"`
	ActualOutcome   string           `json:"actual_outcome"`
}
